File: app/src/main/java/vn/tiki/sample/cart/CartMenuItemView.java
package vn.tiki.sample.cart;

import android.content.Context;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.util.AttributeSet;
import android.widget.FrameLayout;
import android.widget.ProgressBar;
import android.widget.TextView;
import butterknife.BindView;
import butterknife.ButterKnife;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.disposables.Disposable;
import ix.Ix;
import javax.inject.Inject;
import vn.tiki.daggers.Daggers;
import vn.tiki.sample.R;
import vn.tiki.sample.entity.CartItem;
import vn.tiki.sample.repository.CartRepository;

public class CartMenuItemView extends FrameLayout {
  @BindView(R.id.tvItemCount) TextView tvItemCount;
  @BindView(R.id.vProgress) ProgressBar vProgress;

  @Inject CartRepository cartRepository;
  private Disposable disposable;

  public CartMenuItemView(
      @NonNull Context context,
      @Nullable AttributeSet attrs) {
    super(context, attrs);
    inflate(context, R.layout.cart_view_menu_item, this);
    ButterKnife.bind(this);
    Daggers.inject(this);
  }

  @Override protected void onAttachedToWindow() {
    super.onAttachedToWindow();
    showLoading();
    disposable = cartRepository.getAllItems()
        .map(cartItems -> Ix.from(cartItems)
            .map(CartItem::getQuantity)
            .reduce((i, i2) -> i + i2)
            .first(0))
        .map(String::valueOf)
        .observeOn(AndroidSchedulers.mainThread())
        .subscribe(text -> {
          hideLoading();
          tvItemCount.setText(text);
        });
  }

  @Override protected void onDetachedFromWindow() {
    super.onDetachedFromWindow();
    if (disposable != null) {
      disposable.dispose();
    }
  }

  private void showLoading() {
    vProgress.setVisibility(VISIBLE);
    tvItemCount.setVisibility(GONE);
  }

  private void hideLoading() {
    vProgress.setVisibility(GONE);
    tvItemCount.setVisibility(VISIBLE);
  }
}

File: sample/src/main/java/vn/tiki/sample/cart/CartMenuItemView.kt
package vn.tiki.sample.cart

import android.content.Context
import android.util.AttributeSet
import android.view.View
import android.widget.FrameLayout
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.disposables.Disposable
import kotlinx.android.synthetic.main.cart_view_menu_item.view.tvItemCount
import kotlinx.android.synthetic.main.cart_view_menu_item.view.vProgress
import vn.tiki.daggers.Daggers
import vn.tiki.sample.R
import vn.tiki.sample.repository.CartRepository
import javax.inject.Inject

class CartMenuItemView(
    context: Context,
    attrs: AttributeSet?) : FrameLayout(context, attrs) {

  @Inject internal lateinit var cartRepository: CartRepository

  private var disposable: Disposable? = null

  init {
    inflate(context, R.layout.cart_view_menu_item, this)
    Daggers.inject(this)
  }

  override fun onAttachedToWindow() {
    super.onAttachedToWindow()
    showLoading()
    disposable = cartRepository.allItems
        .filter { it.isNotEmpty() }
        .map {
          it.map { it.quantity }
              .reduce { i1, i2 -> i1 + i2 }
              .toString()
        }
        .observeOn(AndroidSchedulers.mainThread())
        .subscribe { text ->
          hideLoading()
          tvItemCount.text = text
        }
  }

  override fun onDetachedFromWindow() {
    super.onDetachedFromWindow()
    disposable?.dispose()
  }

  private fun showLoading() {
    vProgress.visibility = View.VISIBLE
    tvItemCount.visibility = View.GONE
  }

  private fun hideLoading() {
    vProgress.visibility = View.GONE
    tvItemCount.visibility = View.VISIBLE
  }
}
